scipy/spatial/transform/_rotation_xp.py

```diff
@@ -0,0 +1,1138 @@
+    return matrix
+
+
+def as_rotvec(quat: Array, degrees: bool = False) -> Array:
+    xp = array_namespace(quat)
+    quat = _quat_canonical(quat)
+    ax_norm = xp_vector_norm(quat[..., :3], axis=-1, keepdims=True, xp=xp)
+    angle = 2 * xp.atan2(ax_norm, quat[..., 3][..., None])
+    small_angle = angle <= 1e-3
+    angle2 = angle**2
+    small_scale = 2 + angle2 / 12 + 7 * angle2**2 / 2880
+    # We need to handle the case where sin(angle/2) is 0 to avoid division by zero. We
+    # use the value of the Taylor series approximation, but non-branching operations
+    # require that we still divide by the sin. Since we do not use the result where the
+    # angle is close to 0, adding one to the sin where we discard the result is safe.
+    div_sin = xp.sin(angle / 2.0) + xp.asarray(small_angle, dtype=angle.dtype)
+    large_scale = angle / div_sin
+    scale = xp.where(small_angle, small_scale, large_scale)
+    degrees = xp.asarray(degrees, device=device(quat))
+    scale = xp.where(degrees, _rad2deg(scale), scale)
+    rotvec = scale * quat[..., :3]
+    return rotvec
+
+
+def as_mrp(quat: Array) -> Array:
+    xp = array_namespace(quat)
+    one = xp.asarray(1.0, device=device(quat), dtype=quat.dtype)
+    sign = xp.where(quat[..., 3, None] < 0, -one, one)
+    denominator = 1.0 + sign * quat[..., 3, None]
+    return sign * quat[..., :3] / denominator
+
+
+def as_euler(quat: Array, seq: str, degrees: bool = False) -> Array:
+    xp = array_namespace(quat)
+
+    # Sanitize the sequence
+    if len(seq) != 3:
+        raise ValueError(f"Expected 3 axes, got {seq}.")
+    intrinsic = re.match(r"^[XYZ]{1,3}$", seq) is not None
+    extrinsic = re.match(r"^[xyz]{1,3}$", seq) is not None
+    if not (intrinsic or extrinsic):
+        raise ValueError(
+            "Expected axes from `seq` to be from ['x', 'y', 'z'] or ['X', 'Y', 'Z'], "
+            f"got {seq}"
+        )
+    if any(seq[i] == seq[i + 1] for i in range(2)):
+        raise ValueError(f"Expected consecutive axes to be different, got {seq}")
+
+    _device = device(quat)
+    axes = [_elementary_basis_index(x) for x in seq.lower()]
+    axes = axes if extrinsic else axes[::-1]
+    i, j, k = axes
+    symmetric = i == k
+    k = 3 - i - j if symmetric else k
+
+    symmetric = xp.asarray(symmetric, device=_device)
+    sign = xp.asarray(
+        (i - j) * (j - k) * (k - i) // 2, dtype=quat.dtype, device=_device
+    )
+    # Permute quaternion elements
+    a = xp.where(symmetric, quat[..., 3], quat[..., 3] - quat[..., j])
+    b = xp.where(symmetric, quat[..., i], quat[..., i] + quat[..., k] * sign)
+    c = xp.where(symmetric, quat[..., j], quat[..., j] + quat[..., 3])
+    d = xp.where(symmetric, quat[..., k] * sign, quat[..., k] * sign - quat[..., i])
+
+    angles = _get_angles(extrinsic, symmetric, sign, np.pi / 2, a, b, c, d)
+    degrees = xp.asarray(degrees, device=_device)
+    angles = xp.where(degrees, _rad2deg(angles), angles)
+    return angles
```

Comment:
```suggestion
    return _rad2deg(angles) if degrees else angles
```
assuming type annotation is correct